project/src/com/google/daggerquery/executor/QueryExecutor.java

```diff
@@ -43,8 +44,26 @@ public static void main(String[] args) {
       }
 
       Query query = new Query(args[0], Arrays.copyOfRange(args, 1, args.length));
-      List<String> result = query.execute(new SourcesLoader().loadBindingGraph());
-      result.forEach(printStream::println);
+      List<BindingGraph> bindingGraphs = new SourcesLoader().loadBindingGraphs();
+
+      ImmutableList.Builder resultBuilder = new ImmutableList.Builder();
+
+      // We assume that we successfully executed a query only if in at least one graph it was executed without fail.
+      String failureReason = "";
+      for (BindingGraph bindingGraph: bindingGraphs) {
+        try {
+          resultBuilder.addAll(query.execute(bindingGraph));
+        } catch (IllegalArgumentException e) {
+          failureReason = e.getMessage();
+        }
+      }
+
+      ImmutableList resultList = resultBuilder.build();
+      if (resultList.isEmpty()) {
+        throw new IllegalArgumentException(failureReason);
+      } else {
+        resultList.forEach(printStream::println);
+      }
     } catch (IllegalArgumentException e) {
       printStream.println("Execution failed. Reason: " + e.getMessage());
     } catch (IOException e) {
```

Comment:
If the ZIP archive contains a file that is not a Binding Graph textproto, BindingGraph.parseFrom() (Line 60 in SourcesLoader.java) will throw a InvalidProtocolBufferException.
https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/Parser.html#parseFrom-java.io.InputStream-
This looks to be a subclass of an IOException, and so will hit this block here.

Can we add a block to catch InvalidProtocolBufferException and let the user know which file was not formatted correctly?

Reply:
Maybe it makes sense to catch this exception in `SourceLoader` and ignore it? What if the user has .textproto files and one non-textproto that appeared in .zip for some reason, then we can still parse the graph from .textproto and ignore all other files.

Reply:
That's a great solution. 

Could you also add an exception in loadBindingGraphs if we find 0 binding graphs?

Reply:
Done ✅ I throw `FileNotFoundException` if `bindingGraphs` list is empty after reading all files.